### PR TITLE
NER and P&C models

### PR DESCRIPTION
**Pull request by ekmb**
1. Added a Tensor version of sklearn classification report with (f1, precision, recall support) 
2. Refactored NER model and moved data stats/preprocessing to DataDescriptor class
3. Added Punctuation and capitalization model
4. added hydra configs to NER and P&C
5. Jenkins tests added

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -114,6 +112,34 @@ pipeline {
...
+    stage('L3: NER') {
+      when {
+        anyOf{
+          branch 'candidate'
+          changeRequest()
+        }
+      }
+      failFast true
+        steps {
+          sh 'cd examples/nlp/token_classification && python ner.py \
```
Comment (VahidooX) [inline Jenkinsfile:124]: Shouldn't we limit the GPU by setting CUDA_VISIBLE_DEVICES here?

File: examples/nlp/token_classification/conf/ner_config.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  train_ds:
+    prefix: train
+    max_seq_length: 128
+    pad_label: 'O'
+    num_samples: -1
+    ignore_extra_tokens: false
+    ignore_start_end: false
+    overwrite_processed_files: false
+    shuffle: true
+    batch_size: 2
```
Comment (VahidooX) [inline examples/nlp/token_classification/conf/ner_config.yaml:26]: Isn't default batch size too low?

File: examples/nlp/token_classification/conf/ner_config.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    drop_last: false
+
+  language_model:
+    pretrained_model_name: bert-base-uncased
+    bert_checkpoint: null
+    bert_config: null
+    tokenizer: nemobert # 'nemobert' or 'sentencepiece'
+    vocab_file: null
+    tokenizer_model: null # only used if tokenizer is sentencepiece
+    do_lower_case: false # true for uncased models, false for cased models, will be set automatically if pre-trained tokenizer model is used
```
Comment (VahidooX) [inline examples/nlp/token_classification/conf/ner_config.yaml:52]: Shouldn't it be True as default bert is uncases?

File: examples/nlp/token_classification/conf/ner_config.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    name: adam
+    lr: 2e-5
+    args:
+      name: auto
+      params:
+        weight_decay: 0.01
+
+    sched:
+      name: WarmupAnnealing
+      iters_per_batch: null # computed at runtime
```
Comment (VahidooX) [inline examples/nlp/token_classification/conf/ner_config.yaml:71]: If it is always computed by us, do we need it there?

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    overwrite_processed_files: false
+    shuffle: true
+    batch_size: 2
+    num_workers:  2
+    pin_memory: false
+    drop_last: false
+
+  validation_ds:
+    prefix: dev
+    max_seq_length: 128
```
Comment (VahidooX) [inline examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:33]: max_seq_length, pad_label, ignore_extra_tokens, ignore_start_end are duplicated in both data loaders config. Is there a better way to have them in one place? For instance, how about moving them into the language_model section? They can be somehow related? Or having a separate section for them like preprocessing or dataset section?

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    activation: 'relu'
+    log_softmax: True
+    use_transformer_init: True
+
+  optim:
+    name: adam
+    lr: 2e-5
+    args:
+      name: auto
+      params:
```
Comment (VahidooX) [inline examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:75]: Was there a plan to flat these sched and optimzier sections?

File: examples/nlp/token_classification/punctuation_capitalization.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+from nemo.collections.nlp.models import PunctuationCapitalizationModel
+from nemo.utils import logging
+
+
+@hydra.main(config_path="conf", config_name="punctuation_capitalization_config")
+def main(cfg: DictConfig) -> None:
+    logging.info(f'Config: {cfg.pretty()}')
+    trainer = pl.Trainer(**cfg.pl.trainer)
+    model = PunctuationCapitalizationModel(cfg.model, trainer=trainer)
```
Comment (VahidooX) [inline examples/nlp/token_classification/punctuation_capitalization.py:27]: Why should we pass trainer to the model? Is it necessary? I see it has None as default value in the NERModel.

File: nemo/collections/nlp/data/punctuation_capitalization_dataset.py [hunk 8]
```diff
@@ -0,0 +1,445 @@
+# Copyright 2020 NVIDIA. All Rights Reserved.
```
Comment (VahidooX) [inline nemo/collections/nlp/data/punctuation_capitalization_dataset.py:1]: This line is a little bit different from may other file.

File: nemo/collections/nlp/data/token_classification/token_classification_dataset.py [hunk 9]
```diff
@@ -187,59 +189,61 @@ class BertTokenClassificationDataset(Dataset):
             the loss_mask,
         ignore_start_end (bool): whether to ignore bos and eos tokens in
             the loss_mask
+        overwrite_processed_files (bool): whether to overwrite the preprocessed files
     """
 
+    @property
+    def output_types(self) -> Optional[Dict[str, NeuralType]]:
```
Comment (VahidooX) [inline nemo/collections/nlp/data/token_classification/token_classification_dataset.py:196]: Are datasets are supposed to have type system?

File: nemo/collections/nlp/data/token_classification/token_classification_dataset.py [hunk 10]
```diff
@@ -187,59 +189,61 @@ class BertTokenClassificationDataset(Dataset):
...
         max_seq_length,
         tokenizer,
+        label_ids,
         num_samples=-1,
         pad_label='O',
-        label_ids=None,
         ignore_extra_tokens=False,
         ignore_start_end=False,
-        use_cache=False,
+        overwrite_processed_files=False,
```
Comment (VahidooX) [inline nemo/collections/nlp/data/token_classification/token_classification_dataset.py:219]: What does it do exactly? Is it the same as use_cache?

File: nemo/collections/nlp/data/token_classification/token_classification_descriptor.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                all_labels, os.path.join(data_dir, mode + '_label_stats.tsv')
+            )
+
+            logging.info(f'Total labels: {total_labels}')
+            logging.info(f'Label frequencies - {label_frequencies}')
+
+            if mode == 'train':
+                class_weights_dict = get_freq_weights(label_frequencies)
+                logging.info(f'Class Weights: {class_weights_dict}')
+                self.class_weights = fill_class_weights(class_weights_dict, max_id)
```
Comment (VahidooX) [inline nemo/collections/nlp/data/token_classification/token_classification_descriptor.py:80]: Should we set default None values for these properties, so that if train was not passed here?

File: nemo/collections/nlp/metrics/classification_report.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    This metric computes number of True Positive, False Negative, and False Positive examples per class.
+    When doing distributed training/evaluation the result of res=ClassificationReport(predictions, labels) calls
+    will be all-reduced between all workers using SUM operations.
+
+    If used with PytorchLightning LightningModule, include TPs, FNs, and FPs inside validation_step results.
+    Then aggregate them at the end of validation epoch to correctly compute validation precesion, recall, f1
+    using get_precision_recall_f1().
+
+    Example:
+        def validation_step(self, batch, batch_idx):
```
Comment (VahidooX) [inline nemo/collections/nlp/metrics/classification_report.py:36]: Nice!

File: nemo/collections/nlp/metrics/classification_report.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                'macro avg', macro_precision, macro_recall, macro_f1, total_examples
+            )
+        )
+        logging.info(
+            '{:20s}   {:8.2f}   {:8.2f}   {:8.2f}   {:8.0f}'.format(
+                'weighted avg', weighted_precision, weighted_recall, weighted_f1, total_examples
+            )
+        )
+
+        if mode == 'macro':
```
Comment (VahidooX) [inline nemo/collections/nlp/metrics/classification_report.py:134]: Do we want to support micro later?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        Initializes BERT Named Entity Recognition model.
+        """
+
+        self.data_desc = TokenClassificationDataDesc(
+            data_dir=cfg.data_dir, modes=["train", "test", "dev"], pad_label=cfg.train_ds.pad_label
+        )
+        self.data_dir = cfg.data_dir
+
+        # TODO add support for sentence_piece tokenizer
+        if cfg.language_model.tokenizer == 'nemobert':
```
Comment (VahidooX) [inline nemo/collections/nlp/models/token_classification/ner_model.py:59]: Don't we have support for switching between tokenizers in the new NeMo?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+            self.tokenizer = NemoBertTokenizer(pretrained_model=cfg.language_model.pretrained_model_name)
+        else:
+            raise NotImplementedError()
+
+        super().__init__(cfg=cfg, trainer=trainer)
```
Comment (VahidooX) [inline nemo/collections/nlp/models/token_classification/ner_model.py:64]: Is there a reason super.init is after dataset loading?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        )
+
+        if cfg.class_balancing == 'weighted_loss':
+            # You may need to increase the number of epochs for convergence when using weighted_loss
+            self.loss = CrossEntropyLoss(weight=self.data_desc.class_weights)
+        else:
+            self.loss = CrossEntropyLoss()
+
+        # setup to track metrics
+        self.class_report = ClassificationReport(self.data_desc.num_classes, label_ids=self.data_desc.label_ids)
```
Comment (VahidooX) [inline nemo/collections/nlp/models/token_classification/ner_model.py:90]: Can we take the max of label_ids as num_classes to drop the first input?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+        """
+        input_ids, input_type_ids, input_mask, loss_mask, subtokens_mask, labels = batch
+        logits = self(input_ids=input_ids, token_type_ids=input_type_ids, attention_mask=input_mask)
+        val_loss = self.loss(logits=logits, labels=labels, loss_mask=loss_mask)
+
+        subtokens_mask = subtokens_mask > 0.5
+
+        preds = torch.argmax(logits, axis=-1)[subtokens_mask]
+        labels = labels[subtokens_mask]
+        tp, fp, fn = self.class_report(preds, labels)
```
Comment (VahidooX) [inline nemo/collections/nlp/models/token_classification/ner_model.py:128]: How about classification_report instaed of class_report?

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    def setup_test_data(self, test_data_config: Optional[DictConfig]):
+        self._test_dl = self.__setup_dataloader(cfg=test_data_config)
+
+    def _setup_dataloader_from_config(self, cfg: DictConfig):
+        text_file = os.path.join(self.data_dir, 'text_' + cfg.prefix + '.txt')
+        label_file = os.path.join(self.data_dir, 'labels_' + cfg.prefix + '.txt')
+
+        if not (os.path.exists(text_file) and os.path.exists(label_file)):
+            raise FileNotFoundError(
+                f'{text_file} or {label_file} not found. The data should be splitted into 2 files: text.txt and \
```
Comment (VahidooX) [inline nemo/collections/nlp/models/token_classification/ner_model.py:170]: Can we have this comments in the example file too to help the user understand the format of the files?

File: Jenkinsfile (hunk 1 above)
Comment (ekmb) [reply Jenkinsfile:124]: I thought CUDA_VISIBLE_DEVICES doesn't work with PL and we can explicitly specify gpu in trainer. Let's leave it as is for now, we. would need to refactor tests anyway

File: examples/nlp/token_classification/conf/ner_config.yaml (hunk 2 above)
Comment (ekmb) [reply examples/nlp/token_classification/conf/ner_config.yaml:26]: it's just for debugging purposes, will increase

File: Jenkinsfile (hunk 1 above)
Comment (VahidooX) [reply Jenkinsfile:124]: I think Oleksii mentioned in another PR and we should have it when we have parallel tests to isolate them.

File: examples/nlp/token_classification/conf/ner_config.yaml (hunk 3 above)
Comment (ekmb) [reply examples/nlp/token_classification/conf/ner_config.yaml:52]: true is fine, with 'bert-base-uncased' and similar pretrained uncased models, tokenizer automatically does lower casing.

File: examples/nlp/token_classification/conf/ner_config.yaml (hunk 4 above)
Comment (ekmb) [reply examples/nlp/token_classification/conf/ner_config.yaml:71]: @ericharper

File: Jenkinsfile (hunk 1 above)
Comment (VahidooX) [reply Jenkinsfile:124]: What happens when two test are run on single gpu? Can pt handle such cases? We can now leave it like this for now and check again later.

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml (hunk 5 above)
Comment (ekmb) [reply examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:33]: @ericharper but let me leave this as is for now, I followed text_classification config example, if we decided on a different config for nlp, let's do this in a separate PR

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 14 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/ner_model.py:59]: I'm going to port it in a separate PR

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 15 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/ner_model.py:64]: yes, because during init we're setting up dataloaders that require self.tokenizer

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 16 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/ner_model.py:90]: label_ids is optional for tasks that have int labels it won't be provided

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 17 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/ner_model.py:128]: done

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 18 above)
Comment (ekmb) [reply nemo/collections/nlp/models/token_classification/ner_model.py:170]: That's a good point. This is out of the scope of this PR and this decision should be made across the whole collection since every task requires a specific format described in the docs.

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 15 above)
Comment (VahidooX) [reply nemo/collections/nlp/models/token_classification/ner_model.py:64]: This is the first time I saw something like this! We should put a note somewhere for the developers and users.

File: nemo/collections/nlp/metrics/classification_report.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+from pytorch_lightning.metrics import TensorMetric
+
+from nemo.utils import logging
+
+__all__ = ['ClassificationReport']
+
+
+class ClassificationReport(TensorMetric):
+    """
+    This metric computes number of True Positive, False Negative, and False Positive examples per class.
```
Comment (okuchaiev) [inline nemo/collections/nlp/metrics/classification_report.py:27]: This looks great! Thanks @ekmb
